llvm/unittests/IR/ConstantRangeListTest.cpp

```diff
@@ -94,4 +94,122 @@ TEST_F(ConstantRangeListTest, Insert) {
+              GetCRL({{AP0, AP4}, {AP8, AP12}, {AP16, AP18}}));
+
+  EXPECT_TRUE(CRL.unionWith(GetCRL({{APN2, AP2}})) ==
+              GetCRL({{APN2, AP4}, {AP8, AP12}}));
+  EXPECT_TRUE(CRL.unionWith(GetCRL({{AP2, AP6}})) ==
+              GetCRL({{AP0, AP6}, {AP8, AP12}}));
+  EXPECT_TRUE(CRL.unionWith(GetCRL({{AP10, AP16}})) ==
+              GetCRL({{AP0, AP4}, {AP8, AP16}}));
+
+  EXPECT_TRUE(CRL.unionWith(GetCRL({{APN2, AP10}})) == GetCRL({{APN2, AP12}}));
+  EXPECT_TRUE(CRL.unionWith(GetCRL({{AP2, AP10}})) == GetCRL({{AP0, AP12}}));
+  EXPECT_TRUE(CRL.unionWith(GetCRL({{AP4, AP16}})) == GetCRL({{AP0, AP16}}));
+  EXPECT_TRUE(CRL.unionWith(GetCRL({{APN2, AP16}})) == GetCRL({{APN2, AP16}}));
+}
+
+TEST_F(ConstantRangeListTest, Intersect) {
+  APInt APN2 = APInt(64, -2, /*isSigned=*/true);
+  APInt AP0 = APInt(64, 0, /*isSigned=*/true);
+  APInt AP2 = APInt(64, 2, /*isSigned=*/true);
+  APInt AP4 = APInt(64, 4, /*isSigned=*/true);
+  APInt AP6 = APInt(64, 6, /*isSigned=*/true);
+  APInt AP7 = APInt(64, 7, /*isSigned=*/true);
+  APInt AP8 = APInt(64, 8, /*isSigned=*/true);
+  APInt AP10 = APInt(64, 10, /*isSigned=*/true);
+  APInt AP11 = APInt(64, 11, /*isSigned=*/true);
+  APInt AP12 = APInt(64, 12, /*isSigned=*/true);
+  APInt AP16 = APInt(64, 16, /*isSigned=*/true);
+  ConstantRangeList CRL = GetCRL({{AP0, AP4}, {AP8, AP12}});
+
+  // No intersection.
+  ConstantRangeList Empty;
+  EXPECT_TRUE(CRL.intersectWith(Empty) == Empty);
+  EXPECT_TRUE(Empty.intersectWith(CRL) == Empty);
+
+  EXPECT_TRUE(CRL.intersectWith(GetCRL({{APN2, AP0}})) == Empty);
+  EXPECT_TRUE(CRL.intersectWith(GetCRL({{AP6, AP8}})) == Empty);
+  EXPECT_TRUE(CRL.intersectWith(GetCRL({{AP12, AP16}})) == Empty);
+
+  // Single intersect range.
+  EXPECT_TRUE(CRL.intersectWith(GetCRL({{APN2, AP2}})) == GetCRL({{AP0, AP2}}));
+  EXPECT_TRUE(CRL.intersectWith(GetCRL({{APN2, AP6}})) == GetCRL({{AP0, AP4}}));
+  EXPECT_TRUE(CRL.intersectWith(GetCRL({{AP2, AP4}})) == GetCRL({{AP2, AP4}}));
+  EXPECT_TRUE(CRL.intersectWith(GetCRL({{AP2, AP6}})) == GetCRL({{AP2, AP4}}));
+  EXPECT_TRUE(CRL.intersectWith(GetCRL({{AP6, AP10}})) ==
+              GetCRL({{AP8, AP10}}));
+  EXPECT_TRUE(CRL.intersectWith(GetCRL({{AP6, AP16}})) ==
+              GetCRL({{AP8, AP12}}));
+  EXPECT_TRUE(CRL.intersectWith(GetCRL({{AP10, AP12}})) ==
+              GetCRL({{AP10, AP12}}));
+  EXPECT_TRUE(CRL.intersectWith(GetCRL({{AP10, AP16}})) ==
+              GetCRL({{AP10, AP12}}));
+
+  // Multiple intersect ranges.
+  EXPECT_TRUE(CRL.intersectWith(GetCRL({{APN2, AP10}})) ==
+              GetCRL({{AP0, AP4}, {AP8, AP10}}));
+  EXPECT_TRUE(CRL.intersectWith(GetCRL({{APN2, AP16}})) == CRL);
+  EXPECT_TRUE(CRL.intersectWith(GetCRL({{AP2, AP10}})) ==
+              GetCRL({{AP2, AP4}, {AP8, AP10}}));
+  EXPECT_TRUE(CRL.intersectWith(GetCRL({{AP2, AP16}})) ==
+              GetCRL({{AP2, AP4}, {AP8, AP12}}));
+  EXPECT_TRUE(CRL.intersectWith(GetCRL({{{APN2, AP2}, {AP6, AP10}}})) ==
```

Comment:
nit: for `GetCRL({{{APN2, AP2}, {AP6, AP10}}})`

only need two `{` and not three?

Reply:
Nice catch! Done.